service/service.go

```diff
@@ -170,59 +173,96 @@ func NewService(cfg *Config, datadir string) (*Service, error) {
 		minMemoryLayer:  uint(minMemoryLayer),
 		genesis:         genesis,
 		datadir:         datadir,
-		executingRounds: make(map[string]*round),
-		errChan:         make(chan error, 10),
+		executingRounds: make(map[string]struct{}),
 		privKey:         privateKey,
 		PubKey:          privateKey.Public().(ed25519.PublicKey),
 	}
+
 	log.Info("Service public key: %x", s.PubKey)
 
 	return s, nil
 }
 
-func (s *Service) loop(ctx context.Context) {
-	var executingRounds errgroup.Group
-	defer executingRounds.Wait()
+type roundResult struct {
+	round *round
+	err   error
+}
+
+func (s *Service) loop(ctx context.Context, roundsToResume []*round, cmds <-chan Command) {
+	var eg errgroup.Group
+	defer eg.Wait()
+
+	logger := log.AppLog.WithName("worker")
+	ctx = logging.NewContext(ctx, logger)
+
+	roundResults := make(chan roundResult, 1)
+
+	// Resume recovered rounds
+	for _, round := range roundsToResume {
+		round := round
+		s.executingRounds[round.ID] = struct{}{}
+		end := s.roundEndTime(round)
+		eg.Go(func() error {
+			err := round.recoverExecution(ctx, round.stateCache.Execution, end)
+			roundResults <- roundResult{round: round, err: err}
+			return nil
+		})
+	}
+
+	timer := s.scheduleRound(ctx, s.openRound)
 
 	for {
-		s.openRoundMutex.RLock()
-		epoch := s.openRound.Epoch()
-		s.openRoundMutex.RUnlock()
-
-		start := s.genesis.Add(s.cfg.EpochDuration * time.Duration(epoch)).Add(s.cfg.PhaseShift)
-		waitTime := time.Until(start)
-		timer := time.After(waitTime)
-		if waitTime > 0 {
-			log.Info("Round %v waiting for execution to start for %v", s.openRoundID(), waitTime)
-		}
 		select {
+		case cmd := <-cmds:
+			cmd(s)
+
+		case result := <-roundResults:
+			if result.err == nil {
+				broadcaster := s.broadcaster
+				go broadcastProof(s, result.round, result.round.execution, broadcaster)
```

Comment:
use eg?

Reply:
I feel like `broadcasProof` should be a method of `round` shouldn't it? I'm also with @countvonzero here.

"Fire and forget" doesn't go well with a graceful shutdown. `boradcastProof` isn't guaranteed to return in a timely manner. It retries broadcasting possibly for many times (config parameter) and has no way of returning early, e.g. when the timer of the round expires.

Reply:
I'm going to remove broadcaster soon in PR for #168

Reply:
the issue's description is empty. why removing broadcaster? 
i think it's better to do the right thing now instead of relying on future implementation.
the point here is to manage shutdown withe errgroup and there is really no good reason not to do it here.

Reply:
It's just a subtask auto-created from https://github.com/spacemeshos/pm/issues/173. I'm implementing it right now.

Reply:
if you insist, i am ok with it.

> There is no point in fixing it

but it's literally doing `eg.Go()` instead of `go`....

Reply:
It's not _that_ simple because it could block shutdown for a very long time (the default is 500 minutes) if broadcasting fails as there is no cancellation mechanism implemented in `broadcastProof()`.